packages/nimble-components/src/table/specs/cell-state-when-scrolling-hld.md

```diff
@@ -0,0 +1,111 @@
+(**Open Question: Pick one of the following two approaches**)  
+**Option 1:**  
+Fire a `CustomEvent` called `cell-blur` on `cell.cellContentContainer` ([prototype](https://github.com/ni/nimble/commit/57a85347f86aa55c416671f135c484ee96701416#diff-8d5022af0a348aba9a14ea0ef27956d08ebbdb2342cb5dc21d0dde4b4b653decR94)). This allows `TableColumn` implementations to handle the event directly from their FAST template ([prototype](https://github.com/ni/nimble/commit/57a85347f86aa55c416671f135c484ee96701416#diff-9ee1b8660e08ff4f9e24d99f2f70000fc2760b07e9ae00d60aa613fb12f2def5R19)). The event handler can find the control in the cell (via `querySelector`, `firstElementChild`, or similar), and call `blur()` on it ([prototype](https://github.com/ni/nimble/commit/57a85347f86aa55c416671f135c484ee96701416#diff-9ee1b8660e08ff4f9e24d99f2f70000fc2760b07e9ae00d60aa613fb12f2def5R6))  
+**Option 2:**  
+Declare a new optional abstract method `onBeforeFocusedCellRecycled(cell: TableCell)` on the `TableColumn` abstract class ([prototype](https://github.com/ni/nimble/commit/57a85347f86aa55c416671f135c484ee96701416#diff-c042ed462074b89ce3df3dd318a82e80a9d096292a3caf101307704b35035fb3R84)). Implementations can `blur()` the controls in their cells similarly to Option 1 (find the control starting from `cell.cellContentContainer`) ([prototype](https://github.com/ni/nimble/commit/57a85347f86aa55c416671f135c484ee96701416#diff-5058c96c2bc55c407497e6c95b298d71299edba7119549c353cd6985500f135dR36)).  
+In order to call that method, the `TableCell` will raise a `cell-blur` event that will be handled by `TableRow`, which has access to the column to call `column.onBeforeFocusedCellRecycled(cell)` on it (prototype: [cell raises event](https://github.com/ni/nimble/commit/57a85347f86aa55c416671f135c484ee96701416#diff-8d5022af0a348aba9a14ea0ef27956d08ebbdb2342cb5dc21d0dde4b4b653decR100), [row event binding](https://github.com/ni/nimble/commit/57a85347f86aa55c416671f135c484ee96701416#diff-bf998763585210ba31b88a11d6ffe26b09f6ee2860c05c20c3c2a3d6c3cd01d8R21)).
+
+### Close Focused Action Menus
+
+The table will handle this internally (in the `Virtualizer` class, via `handleVirtualizerChange()`), without affecting the TableColumn public API.
+
+We want to close the action menu via the associated `MenuButton`, which allows the rest of the table logic dealing with action menus to get called normally.
+In this case, `table.shadowRoot.activeElement` will be null (since the action menus are slotted in), but `document.activeElement` will be a Nimble `MenuItem`. (We can also doublecheck that `table.contains(document.activeElement)` before proceeding.) From the MenuItem, we can walk up the `parentElement` hierarchy to get to the Nimble `Menu`. Then we can walk up the `assignedSlot` hierarchy until we get to the TableCell action menu slot (in which case `slot.parentElement` is a Nimble `TableCell`). At that point, we can get the `MenuButton` for the cell, and call `open = false` on it.
```

Comment:
Rather than having to walk the DOM tree & make assumptions about the focused element, an alternative is to use the information we already have in the table & table-row about which menu is open.

Specifically:
- the table has an observable property named `openActionMenuRecordId`
- the table row has `menuOpen` and `currentActionMenuColumn`

Without thinking through the exact code, my initial preference is to use the information we have already (we might find we want to make some slight changes to reset some values back to `undefined` when menus are closed) to programmatically close the open menu without walking the DOM.

Reply:
I took a stab at this in the [updated prototype](https://github.com/ni/nimble/compare/@ni/nimble-components_v18.6.1...table-cell-state-custom-elements-1#diff-e2bfd4eda0a3c89f54b4e09624e6cfd5a68ea2f14c0e79f717eacce38ec5982bR132), but the code still doesn't seem optimal.
- `openActionMenuRecordId` to `TableRow` - I ended up doing ``viewport.querySelector(`nimble-table-row[record-id="${this.table.openActionMenuRecordId}"]`)``, not sure if there's a more straightforward way
- From `currentActionMenuColumn` I don't think we can easily get to a `TableCell` - given a row, we can get all the cells (via DOM APIs) and find the one with `menuOpen`, so the prototype does that currently.

For now the HLD mentions both options. The prototype also has my original code that walks the DOM still there but commented out, for comparison. Not sure if we need to nail down all the details here (in this HLD) or not.

Reply:
As mentioned above I think we can nail down the details more here with the actual implementation (separate from the HLD), so marking Resolved.